Relationship tools & stronger Unfollow feature & new `interact_by_URL` feature & lots of good revisions

uluQulu opened the pull request:
#### After getting too many requests from bunch of people last times, I decided to write relationship tools.
+ And while writing it, I felt the need to modify Unfollow feature.
+ Out of the plan, with the request of a user I prepared a new feature: `interact_by_URL` and included it in this commit series
+ Wrote a `progress_tracker` from scratch to improve user awareness (_it's even better than `tqdm` module :P_).  
    _For now, `progress_tracker` is integrated only to a pair of relationship tools._
+ Yes, it did not end up, I also added a new `highlight_print` definition to have a tasteful console output.

_Those are the notable changes that have happened so far._

Now I will tell briefly about the changes.
(_to save some lines and for much better view I've attached their links from README to see and read the details in full_)

#### [Relationship tools](https://github.com/uluQulu/InstaPy#relationship-tools)- the fresh new features come with it:
+ [Grab `Followers` of a user](https://github.com/uluQulu/InstaPy#grab-followers-of-a-user):
```python
grab_followers(username="Maria", amount="full", live_match=True, store_locally=True)
```
+ [Grab `Following` of a user](https://github.com/uluQulu/InstaPy#grab-following-of-a-user):
```python
grab_following(username="Maria", amount=900, live_match=True, store_locally=True)
```
+ [Pick `Unfollowers` of a user](https://github.com/uluQulu/InstaPy#pick-unfollowers-of-a-user):
```python
pick_unfollowers(username="Maria", compare_by="month", compare_track="median", live_match=True, store_locally=True, print_out=True)
```
+ [Pick `Nonfollowers` of a user](https://github.com/uluQulu/InstaPy#pick-nonfollowers-of-a-user):
```python
pick_nonfollowers(username="Maria", live_match=True, store_locally=True)
```
+ [Pick `Fans` of a user](https://github.com/uluQulu/InstaPy#pick-fans-of-a-user):
```python
pick_fans(username="Maria", live_match=True, store_locally=True)
```
+ [Pick `Mutual Following` of a user](https://github.com/uluQulu/InstaPy#pick-mutual-following-of-a-user):
```python
pick_mutual_following(username="Maria", live_match=True, store_locally=True)
```

#### After _revisions_, _unification_ and _enhancement_- [the fresh new **Unfollow** feature](https://github.com/uluQulu/InstaPy#unfollowing) has been formed:
```python
unfollow_users(amount=10,
               customList=(False, [], "all"),   #this is a new method
               InstapyFollowed=(False, "all"),   #previously called "onlyInstapyFollowed"
               nonFollowers=False,   #previously called "onlyNotFollowMe"
               allFollowing=False,   #previously existed but did not have a parameter, now it is explicitly a new parameter
               style="FIFO",
               unfollow_after=None,
               sleep_delay=600)
```

> Shortly, by now, `style` and `unfollow_after` parameters both are supported by all of the 4 available unfollow methods. And `customList` & `InstapyFollowed` methods now have a track to unfollow with- `"all"` or `"nonfollowers"`.

#### [Interact on posts at given URLs](https://github.com/uluQulu/InstaPy#interact-on-posts-at-given-urls)
```python
interact_by_URL(urls=["some/URL/1", "some/URL/2" "other/URL"], randomize=True, interact=True)
```

#### Please have a look at the documentation which addresses most of the questions that may raise in mind.

+ **Quick tip**: all of the relationship tools works with any username provided, it can be yours own or a username of any non-private account (_if your account is private, it will work, too_).
+ **Note**: All of the code has been specifically chosen and tested for platform and python (_version_) independent environments.

Those standalone data grabber tools will make lots of new implementations available. (_e.g. finding `Ghosts`_)
> Note that I did use the already existing functionality of data scraping which resided in the unfollow feature's priorly called `onlyNotFollowMe` condition and added improvements to it to form latest grabbers.

* BTW @timgrossmann, your issue is 50% complete now, the one you have requested to write scrolling parts with graphql, and now, all of the _followers_ and _following_ information of **any** user can be done by these tools in 5 times faster speed.

Lastly, it can be said that these exclusive tools is gonna make **InstaPy** rich and unique.

If you have any questions, please leave a comment to start a discussion.
Best regards, 
Şəhriyar.

uluQulu commented:
Hey @sionking 
The change you have done in [this commit](https://github.com/timgrossmann/InstaPy/commit/dcb07471d417a51d0c223f7cebcf5c47281aa7e6#commitcomment-29358601) is for good.
Actually I realized it only after running lots of unfollow tests and I there I was wondering why it does not unfollow lots of users without a notice.
And then after a while I found out the source of the problem, it was an innocent `dont_include`.
And for one particular condition (_which unfollows everybody, now `allFollowing` method_) it had no any handle, it was just continuing to the next user without saying that it is in a "_whitelist_" (_of `dont_include`_). Now I added a logger message to that place so that users understand why it skipps some users.

But while merging the changes with the latest files, I saw that there (although, in other unfollow method-now `InstapyFollowed`) it now removes the users found in `dont_include`. Which is an unwanted behavior. So I reverted back that change. (_that's the only thing I have reverted back with this PR, everything else is either improved or revised_)

The problem relies in an implementation consideration:
+ If `blacklist` is enabed, in the very start it will dump all of the users in blacklist to the `dont_include` container. 
It means, now, `dont_include` not only keeps the list of friends, but also the users who were previously interacted WHO is now part of a blacklist campaign and part of the `dont_include` variable.

The point I underline is, that mechanism have to be rewritten.
Cos if you just delete every user `dont_include` while blacklist is enabled, it will delete those data of your friends from the CSV files (_This is fine, cos a friend don't need to be in that CSV_) AND will also delete those unrelated guys from CSV.

Hope you got my point.
![dont_include](https://user-images.githubusercontent.com/33269249/41464614-031cb77c-70ac-11e8-9515-dc1990d3d8bc.png)


Cheers

SammyHam commented:
@uluQulu dude you rock! (y)

sionking commented:
@uluQulu I agree blacklist feature not fit well in instapy design, I actually use it as a logger for actions :) but not for dont_include.
I got your point anyway, but still dont_include users "whitelist" should never be unfollowed.

sionking commented:
@uluQulu 
Hi, I really like what you did here, I for long time only interact by first scraping the followers of a user by graphql. but I do have some warning, don't scrap the followers with the same user you do interactions, this is a red flag.

I think we have major issues we don't address, beside "targeting" profiles:
1. detectable, I am working on a mouse movement with human behavior , I can show you that, since I can't find time to start/finish it, I have the working mechanism but no time to integrate it into instapy.
2.  limits limits limits, I have modifieded version of follow/unfollow/like that also check if the action really done, I think most of the ppl here are not aware that 5% of the action not even happening (automaticly checking the like in the liked url by a different user)

uluQulu commented:
Thanks really, all of you guys!
I really appreciate your congenial reviews.

@timgrossmann, I think, InstaPy is more than a useful program, it's also a great programming school and you're a great teacher!
The reason I could accomplish to write all these pieces here and there is mainly because of having the program design being so clean, source code structure resilient and having the documentation prepared with brilliance.
I have used tons of software to this day and I can surely say that there are bunch of commercial software missing those advantages.
Now this truly useful and in-demand project, also helps to shift up the open source software movement having too many people involved into it. 

And in my perspective, I think, it will keep helping many guys to get started with programming and be thankful to InstaPy.

Regards.

uluQulu commented:
#### After the introduction of new tools, I will talk a bit about the details

I have to tell them cos currently it's only in my mind, and it must be shared for now and the future for those mechanisms got understood clearly.

_I assume that you have seen the documentation of the tools_

#### `live_match`
+ What is live match? 
The reason I have prepared those tools was mainly cos having people could easily access them whenever wanted.  
Fine, but when I went ahead and started to re-integrate it with **Unfollow** feature's `nonFollowers` method, I noticed one thing.
Imagine a user has multiple a **quickstart** with multiple unfollow activities declared at several separate positions which is pretty normal. And all (_or some- more than one_) of them are using `nonFollowers` method.
```python
###quickstart script
##some code
unfollow_users(amount=50, nonFollowers=True, style="RANDOM", unfollow_after=24*30*60, sleep_delay=500)
##some other code
#~30 minutes passed and now will unfollow, again with `nonFollowers` method
unfollow_users(amount=66, nonFollowers=True, style="LIFO", unfollow_after=38*60*60, sleep_delay=655)
###etc.
```
There, I noticed that in each **Unfollow** activity with `nonFollowers` method, it will _grab_ `followers` and `following` data **from scratch** to find `nonfollowers`.
Here the trouble lies, since that activity is request hungry and is not welcomed for queries in short frequencies. There should be a mechanism to smartly handle it.
Then I started to think about **LIVE MATCH**.
+ What it is for? 
Great now we realized why it is needed.
Now. there is a in-class variable
```python
self.relationship_data = {username:{"all_following":[], "all_followers":[]}}
```
which holds the **sessional** relationship data of any user after it's data has been grabbed in `"full"` range.
So, in the very first **Unfollow** activity, it will go and grab that relationship data in `"full"` range.
And then, when the second **Unfollow** activity starts, what it will do is done by `LIVE MATCH` algorithm.
There I explain one more thing before it. `IG` stores the relationship data in `FIFO` style which made **LIVE MATCH** work.
The thing is the relationship data retrievement for the Unfollow activities are being happened with the `live_match=False` state, which means in the second Unfollow activity, it will go and grab **only the new usernames** (_if any_) from the server.
The result is, in the second and further Unfollow activities, `nonFollowers` data is being retrieved in a few seconds (_even milliseconds_) rather than some bunch of TEN minutes and spending big amounts of server requests which is not welcomed.

+ How the **LIVE MATCH** algorithm works?
Well, to find out the new users from the server I used a simple logic.
We learnt that `IG` stores data in `FIFO` style;
Now if there will be either a new `follower` or a `following`, it will place in the top of the data.
Great. There, **LIVE MATCH** is gonna do some logic.
It will look for `10` (_I chose it, but can be changed to e.g. `50`_) **old usernames** (_old usernames found in relationship data sessional storage resided after the first query in `"full"` range_) in the **current data** and if it find those `10` people, THERE it stops querying and sums up the **NEW** and **OLD** data to prepare a **final DATA**.
It's smart right?

+ How reliable is **LIVE MATCH**?
When you get data. Imagine user A was not following you (_whom you are following_) and when it follows you back during the first and the second **Unfollow** activity, THEN that user A will stay on top of the data at the server and **LIVE MATCH** will not get that user info.
Now imagine user B was following you and during the first and the second Unfollow activity user B unfollows you. Well this time you during that session that user B will not be marked as a Nonfollower, instead in the next session, it can be evaluated as a nonfollower and added to unfollow list.
This time imagine, you are following user C. Fine, during the first and second Unfollow activity, you go unfollow the user C either with `unfollow` definition or with `unfollow_user` (_standalone unfollower_) definition. Well there I wrote a few lines, which is removing the user C from the `relationship_data` sessional storage. But okay, what if you follow that user C back after these last steps: _"unfollow+remove data"_, well if you follow that user C again, then that user C will appear at the top of the `Following` data located at server. So that LIVE MATCH will grab that user data.

As you see, LIVE MATCH does not pay from accuracy. And if you have concerns, just increase the matching value from `10` to e.g. `50` or even `100`. 😂

+ Why it is existing in all those tools?
I mainly designed it for for Unfollow feature to have a fluent activity flow and save requests.
Although, then I thought users might also would like to use it in case of they would try to do multiple initializations.
And the good thing is it now is supported by all of the query based features. And it is up to you to use it.
Well, for Unfollow feature there might not be any leak in accuracy, but for other features, if you like ultimately lvie data, use `live_match=True` or it's not a priority, disable live_match and use smart matching as I explained above which will help you a lot.
One more time saying, those evaluations are per session. It means if you start a new session, it will always get relationship **LIVE** data in the first data request, and in the next ones, it is up to your `live_match` choice..

Why is this post so long?
+ I had to explain this mechanism with _in_s and _out_s, cos it is the basic bone of the these data grabbing tools

uluQulu commented:
### The second major change happened in UNFOLLOW feature also needs an explanation.

I started to modify it simply cos of the data grabber was dislocated into a separate definition.
Right, while doing it, the other changes came along automatically cos I when I touch one thing I always try to do it in its best.

First of all I should say that, previously, **Unfollow** feature had this structure (_simplified_):
```python
if InstapyFollowed:
    #get data from local file
    #go unfollow from profiles one by one
elif nonFollowers==True:
    #get data from graphql queries
    #go unfollow from profiles one by one
elif allFollowers:
    #get data from dialog box
    #go unfollow right from dialog box
```
I noticed that both `InstaPy` and `nonFollowers` methods **both** are doing unfollows with the same method: 
`#go unfollow from profiles one by one`
Heh, the difference is they get data from different data sources.
The problem is, although they both have the same procedure of unfollowing (_by visiting profiles_) just because of they each own a separate processes of doing the same thing, they make the work for maintainer, implementor very hard. 
Cos every time if there is a new thing, it must be applied to both of them separately and most of time maintainers forget one of those and that's why the same procedure for `InstapyFollowed` is rich but that procedure for `nonFollowers` Is poor.

And yeah I also added a new condition: `customList` and the final story is as below:
```python
if (customList == True or
        InstapyFollowed == True or
            nonFollowers == True):
    #get data according to the method
    #go unfollow from profiles
elif allFollowing == True:
    #get data from dialog box
    #go unfollow from dialog box
```
as you see, now there is only 2 ways of unfollowing:
+ code of doing unfollow by visiting Profiles one by one (_3 methods using this_)
+ code of doing unfollow from Dialog box (_the other 1 method uses this_)

The result is great
+ now every time a maintainer changes something is implementor adds something, it will be much more easier.
+ now code is more clearer, and there are less lines (_otherwise imagine customList also had a separate code block for that.._- it was gonna be 3 duplicates of the same thing)

### Add `style` support to all of them
+ Well previously `style`s were supported only by `InstapyFollowed` **Unfollow** method.
+ And the code of finding nonfollowers for `nonFollowers` method was using `set` which was changing the order of the elements:
```python
nonfollowers = set(allfollowing) - set(allfollowers) - set(dont_include)
```
+ And the last method `allFollowing` was more complicated, it was zipping usernames and buttons together while starting the unfollow loop,.

Now, I changed using `set` to 
```python
#using this approach we can preserve the order of elements to be used with `FIFO`, `LIFO` or `RANDOM` styles
nonfollowers = [user for user in all_following if user not in all_followers]
#uniqify elements
nonfollowers = sorted(set(nonfollowers), key=nonfollowers.index)
```
And for the `allFollowing` method, I pre-zipped them before the loop and added all desired layers then served to the unfollow loop.

##### The result is style is fully functional and working for all 4 Unfollow methods.

### Add `unfollow_after` support for all:
This has costed to do some modifications in `automatedfollowedPool`,
```python
automatedFollowedPool = {"all":[], "eligible":[]}
```
And now it will say:
```powershell
Total 3002 users available to unfollow  ~not found in followedPool.csv 456  |  didn't pass `unfollow_after`: 29
```
**Shortly**:
It will always generate an `eligible` list from `automatedfollowedPool` before starts an unfollow activity.
Then, for `customList`, `nonFollowers` and `allFollowing` methods, if the user exists in that local CSV file, then it will check accordingly:
+ generate an unfollow list (_according to the method_)
+ If user does not exists in the CSV, keep it in generated unfollow list
+ If user exists in the CSV and is not `eligible` to unfollow, then remove it from unfollow list.
+ If user exists in the CSV and is `eligible`, then keep it in the unfollow list.

Plain and simple.
So that after all, `unfollow_after` is fully supported by all of the 4 methods.

I will not go deep into other modifications in **Unfollow** feature.
But there are notable changes to mention, e.g.
+ I revised sleeping mechanism for the Unfollow from profiles procedure,..
+ Also, now `allFollowing` method will always get the accurate amount of users to unfollow, cos it is now integrated to use the other good working definition- `get_users_through_dialog`.
+ Well yes, there is `track`s for `customList` and `InstapyFollowed` methods, etc.

If you still wonder something, please ask it.

uluQulu commented:
There are too many blocks that need explanation.
But lastly I will quote important ones.
#### 1. There will be a new file written in logs folder called- **graphql_queries.json**i
Its structure is simple:
```javascript
{"userA": {"28-05-2018": {"sc_rolled": 7}, "29-05-2018": {"sc_rolled": 569}}, "userB":{"10-06-2018": {"sc_rolled": 300}}}
```
It will store the overall scroll count of graphql queries.
Currently its use case is only for personal analysis, e.g. you can view any time that how many times you have queried through graphql from any account, any day.
But in future it will help to generate a such as graphql **query limiter** function.
#### 2. Data storing from the relationship tools
I have explained all of the information in the documentation on how and where the data will be stored, but will tell a few more bits now:
I used the filetype of **json**, mainly cos those data files will be viewed by users itself so it should be more userfriendly type like **json** rather than one liner plain `TXTS or `CSV`.
And loading JSON is more easier. And to prevent corruption, I added `interruption_handler` to all of those writes.
Also, if the filename is gonna be written already exists in that directory, e.g. a data from the same date, with the same amounts. Then it will write with its index:
`01-06-2018~full~6681.json`
`01-06-2018~full~6681(1).json`
You may think that the files with the same amount can be identical, yeah and they can also not be identical, e.g. some users are replaced with others but the amount is still the same, so, indexing rather than overwriting old file is a good choice.
Also there is a layer that do check: if the data gonna be written is identical with the data in the previous query stored in the relationship_data in-class variable, then it will not store that data locally again.
```python
if len(all_followers) > 0:
    if (store_locally == True and 
            relationship_data[username]["all_followers"] != all_followers):
        store_followers_data(username, grab, all_followers, logger, logfolder)
    elif store_locally == True:
        print('')
        logger.info("The `Followers` data is identical with the data in previous query  ~not storing the file again")
```
#### It was difficult to get this function ready: `load_followers_data`
It mainly is doing the work of `compare_by` and `compare_track` to find unfollowers of a user and that definition have to return one optimal file.
I did manage to get it work as desired and I THINK capture lots of lines to explain its mechanism, I encourage you to read it's source code to see the story.
>**Hint**: `glob` module helped me a lot in there

>**NOTICE**: Keeping the filenames as they are written is the KEY to have a good working **unfollower picker**. If you rename those files in **unfollowers** folder, the result will be: that function- `load_followers_data` will not be able find those files in there and will say you: "_Hey generate some followers to find your unfollowers in the future!_"

#### Progress Tracker
This is more fun.
![progresstracker2](https://user-images.githubusercontent.com/33269249/41496433-48a78f74-7151-11e8-9ce1-7c3b74349ed9.gif)
I wanted a progress tracker cos I didn't like waiting for some minutes while grabbing relationship data and console was all empty staring at me.
So with progress tracker user will know how many seconds/minutes has left for data grabbing finish, etc.
>Idea: In future it can be integrated into other long time waiting activities easily cos it needs only simple values,
```python
progress_tracker(current_value, highest_value, initial_time, logger)
```
Also, it is just 36 lines of code and is better than progress bar modules like `tqdm` which needs a pip install :P
Quick Tip: I used `sys.stdout.write("progress")` to be able to produce a python (_version_) independent code. Otherwise, I had to use `print("progess", end='\r')` for python 3 and `print "progress",` for python 2.

#### `highlight_print`
It has been rather easy, 
```python
output_len = 28+len(username)+3+len(message)
```
It is the length of a typical logger output
```powershell
INFO [2018-06-16 10:51:02] [Mosquito]  --> Total people unfollowed : 36
```
e.g. for this logger message, it will be
```python
28 = len("INFO [2018-06-16 10:51:02] [") 
len(username) = len("Mosquito")
3 = len("]  ")
len(message) = len("--> Total people unfollowed : 36")

```
so it will produce a good looking character set for that `output_len` and write away in console with.
Please see it's source code for more better understanding. I added this kind of highlighted print only to "feature" startings, session "initialization" and "end" and "login" only.

### There are several improvements in the already existed data grabbing block
E.g. now it has a break time handler
```python
#take breaks gradually
if sc_rolled > 91:
    print('\n')
    logger.info("Queried too much! ~ sleeping a bit :>")
    sleep(600)
    sc_rolled = 0
```
### I will not talk about `interact_by_URL` :D
The other changes are more serious and this particular feature is gonna do simply regular activity like its other interaction based friends, but a bit more improved than them.

I think I have covered every important block. Good, my responsibility is done guys, it's up to you now to decide on.

Also I will have busy weeks ahead having an exam this month and then after it I will possibly not be able to come back to this place for a long time.
That's why I encourage you to take considerations and ask me whatever you wonder that I can help of as long as I am here.

Cheers.

uluQulu commented:
There are too many blocks that need explanation.
But lastly I will quote important ones.
#### 1. There will be a new file written in logs folder called- **graphql_queries.json**i
Its structure is simple:
```javascript
{"userA": {"28-05-2018": {"sc_rolled": 7}, "29-05-2018": {"sc_rolled": 569}}, "userB":{"10-06-2018": {"sc_rolled": 300}}}
```
It will store the overall scroll count of graphql queries.
Currently its use case is only for personal analysis, e.g. you can view any time that how many times you have queried through graphql from any account, any day.
But in future it will help to generate a such as graphql **query limiter** function.
#### 2. Data storing from the relationship tools
I have explained all of the information in the documentation on how and where the data will be stored, but will tell a few more bits now:
I used the filetype of **json**, mainly cos those data files will be viewed by users itself so it should be more userfriendly type like **json** rather than one liner plain `TXTS or `CSV`.
And loading JSON is more easier. And to prevent corruption, I added `interruption_handler` to all of those writes.
Also, if the filename is gonna be written already exists in that directory, e.g. a data from the same date, with the same amounts. Then it will write with its index:
`01-06-2018~full~6681.json`
`01-06-2018~full~6681(1).json`
You may think that the files with the same amount can be identical, yeah and they can also not be identical, e.g. some users are replaced with others but the amount is still the same, so, indexing rather than overwriting old file is a good choice.
Also there is a layer that do check: if the data gonna be written is identical with the data in the previous query stored in the relationship_data in-class variable, then it will not store that data locally again.
```python
if len(all_followers) > 0:
    if (store_locally == True and 
            relationship_data[username]["all_followers"] != all_followers):
        store_followers_data(username, grab, all_followers, logger, logfolder)
    elif store_locally == True:
        print('')
        logger.info("The `Followers` data is identical with the data in previous query  ~not storing the file again")
```
#### It was difficult to get this function ready: `load_followers_data`
It mainly is doing the work of `compare_by` and `compare_track` to find unfollowers of a user and that definition have to return one optimal file.
I did manage to get it work as desired and I THINK capture lots of lines to explain its mechanism, I encourage you to read it's source code to see the story.
>**Hint**: `glob` module helped me a lot in there

>**NOTICE**: Keeping the filenames as they are written is the KEY to have a good working **unfollower picker**. If you rename those files in **unfollowers** folder, the result will be: that function- `load_followers_data` will not be able find those files in there and will say you: "_Hey generate some followers to find your unfollowers in the future!_"

#### Progress Tracker
This is more fun.
![progresstracker2](https://user-images.githubusercontent.com/33269249/41496433-48a78f74-7151-11e8-9ce1-7c3b74349ed9.gif)
I wanted a progress tracker cos I didn't like waiting for some minutes while grabbing relationship data and console was all empty staring at me.
So with progress tracker user will know how many seconds/minutes has left for data grabbing finish, etc.
>Idea: In future it can be integrated into other long time waiting activities easily cos it needs only simple values,
```python
progress_tracker(current_value, highest_value, initial_time, logger)
```
Also, it is just 36 lines of code and is better than progress bar modules like `tqdm` which needs a pip install :P
Quick Tip: I used `sys.stdout.write("progress")` to be able to produce a python (_version_) independent code. Otherwise, I had to use `print("progess", end='\r')` for python 3 and `print "progress",` for python 2.

#### `highlight_print`
It has been rather easy, 
```python
output_len = 28+len(username)+3+len(message)
```
It is the length of a typical logger output
```powershell
INFO [2018-06-16 10:51:02] [Mosquito]  --> Total people unfollowed : 36
```
e.g. for this logger message, it will be
```python
28 = len("INFO [2018-06-16 10:51:02] [") 
len(username) = len("Mosquito")
3 = len("]  ")
len(message) = len("--> Total people unfollowed : 36")

```
so it will produce a good looking character set for that `output_len` and write away in console with.
Please see it's source code for more better understanding. I added this kind of highlighted print only to "feature" startings, session "initialization" and "end" and "login" only.

### There are several improvements in the already existed data grabbing block
E.g. now it has a break time handler
```python
#take breaks gradually
if sc_rolled > 91:
    print('\n')
    logger.info("Queried too much! ~ sleeping a bit :>")
    sleep(600)
    sc_rolled = 0
```
### I will not talk about `interact_by_URL` :D
The other changes are more serious and this particular feature is gonna do simply regular activity like its other interaction based friends, but a bit more improved than them.

I think I have covered every important block. Good, my responsibility is done guys, it's up to you now to decide on.

Also I will have busy weeks ahead having an exam this month and then after it I will possibly not be able to come back to this place for a long time.
That's why I encourage you to take considerations and ask me whatever you wonder that I can help of as long as I am here.

Cheers.

uluQulu commented:
Glad that I helped, @xX7 😀
Generally I can understand things easily at first glance but actually, what you are doing looked too complex to me, really, and I think you will produce something very valuable using that sophisticated logic.

Real mouse movement- wow, @sionking sounds crazy!
To be honest I have not thought about it and now I see that possibility which is amazing.
I hope you can finalize those great works and introduce it to the community.

About this toolset, @sionking 
Sure, I have read your words long time ago, on @timgrossmann's Issue regarding the `manual scrolling` **>** `graphql scraping` and honestly, I somehow agreed with your point.

But the reason I have prepared these tools is:
+ I wanted to help people to be able to at least get a **backup** of their relationship data
+ And do it easily through InstaPy which they prob are familiar with

I did not have intention to write all other features at first, but after preparing standalone grabbers, I thought why not to write a an **unfollower**, **nonfollower**, **fans** and **mutual** pickers, etc.

Now people can do it all easily within InstaPy.
And the `live_match=False` functionality is amazing. What it means is,
You can grab the data from servers only once and using the **LIVE MATCH** algorithm, you can get all of those information at once in a few seconds!

_I **recommend** you to read more about **LIVE MATCH**._
>Also, I have given `live_match=True` in defaults, maybe it is better to give it `False`, eh?

**Again, I have to say that these tools must be used with care and NOT overused as any other feature exists.**

#### To your proposal, it is hard to to achieve it:
+ Not everybody has an empty accaunt to spend to querying BUT okay, there can be a mechanism to get a make user for just querying data (_a definition for that_)
+ Fine but the process of getting a PROXY will not be easy. Cos I don't know that easy solution to that. Because there are people from several places of the world and one proxy will not work in other.

>Getting those steps solved together is too difficult and again imagine it is solved, then it will need another instance to get those data at the same time. Which makes 2 InstaPy + 2 Browser instances running at the same time **ALWAYS**- not performance friendly.

#### What is the optimal solution?
Add a good query limiter function. So if user queries `N` amount of times per day (_can be checked in hourly basis, too_), then DO not allow further querying anymore for that day (_or hour_). 
+ And if in the future manual scroll parts wil be merged to these queries then start to use those manual scrollers until new queries opens (_next day or next  hour_).

sionking commented:
@uluQulu 
**Live Match** is an amazing tool, as long as you have FIFO, which is not always the case, I am not sure how they choose to do it but for 95% of times it will be FIFO, but counting on it ? sure for normal use.
I think a better way will be to capture **all followers** of a profile (once with a proxy), then start live match with the existing database of all followers as a starting point.

I am sorry to say I will not take the risk of using **ANY** non human behavior, with my real accounts, scrapping just with graphql is easy to track (or am I wrong?), but you all welcome, maybe it just me that worry to much.
I really can use the tools for many other things.

Can you help me here:
1. I want to grab all followers of A.
2. I want to run over the followers of that user.
3. I want to update the followers since it finished all.

- What should be done ?
- what is the method that keep track of which followers I already visited ?

let's say I have 100 followers of user A stored in a file, now I visited 51, tomorrow I want to visit the next 49, where this pointer "51" is stored ? or am I missing something.

xX7 commented:
@uluQulu Thank you, I hope I will, due to having limited time, because of my job :(

@sionking As far as I know, that graphql call is also done by IG, if you manually scroll the list in the background, via an AJAX call.

An approach to look legit, would be to actually go to the users page, click on the followers button to open the dialog and then do the graphql query. I would not see any method IG would use to detect that this is not a normal behaviour. Generally speaking, IG would need logs and perfect algo's to detect that. EVEN if they would log the behaviour/unusual behaviour, these logs would be HUGHE, looking at the amount of users, browsing IG every day. 

Of course there are tools that I know, that somehow "record" the session and the flow the user takes of your platform. These tools are made to analyse and optimize the frontend afterwards via "A/B" tests, but I highly doubt that these tools are used to detect "bot behaviour". We would also see such a tool in the loaded javascripts.

sionking commented:
Look at IG last mail.
They record every movement.

uluQulu commented:
Sure it is not gonna change easily, @sionking cos it affect the speed of retrievements in the server side devastatingly.
Well, it's just a parameter you can enable live data retrievement with `live_match=True` always and skip the matching algorithm.

I have wrote my thoughts there, there is no any optimal way of applying proxy to all users. 
Sure you can tell users find a proxy and provide it, it can be done but a user who cannot properly write a custom line in quickstart will fairly fail to find a proxy and even understand what is proxy, even if finds, it will be a proxy of the other sie of the world which will result more insecure connection.
Please have a read my post mentioning it above.

These tools are for data grabbing and picking, nothing more nothing less.
You just cannot expect them to store the visited users. It is just out of the scope of this toolset.
What can do is:
+ Grab the data at full or desired amount OR use any picker tool
+ At any of those feature calls, if the `store_locally` is enabled, it will store that data into a file.
(_data retrievement through the inner calls automatically store data, e.g. while getting nonfollowers for Unfollow feature, etc._)

If you like, you can easily write some lines to also store visited users for you and there is a great variable which it can benefit-  `relationship_data` in-class dictionary.

uluQulu commented:
Nice @xX7 
I didn't know about it, although we do not run AJAX calls it is a good news, really.
You're also right about continuous logging, but I think it will be really difficult to generate a native-looking activity.

sionking commented:
@uluQulu 
I personally, keep a number for each user followers, and start from this number
e.g. nike-followers : 500, adidas-followers:1234 ... and the list of followers don't change it just update so the first 1234 in adidas will not be visit again.

Those are really great great tools and really nice written and explained! thanks you for the hard work !

xX7 commented:
@uluQulu yes you can verify it by using the network tool of chrome for example, just scroll through some follower's/following dialog. The thing is, that IG always uses 12 as the count for the pagination... 
Anything above 12 may look suspicious as it is a direct API call and therefore can easily be logged.

On the other hand, I don't really know if this specific endpoint **CAN** be used for other reasons and therefore is somehow "public". If that is the case, they cannot just look for the count=12 as many people would call that API with different variables....

I think the best and safest way to use the graphql for getting user's followers/following etc would be to create a IG account with a proxy and only use that API via a proxy to feed InstaPy...

timgrossmann commented:
@uluQulu Love it! Thank you very much for the detailed explanation here, learned a lot. 🙏 

@sionking I think you're right in the sense that we should make sure to have a "simulated mouse movement". This was one of the ideas in using selenium from the beginning that it is more "human like" than simply using their API 100%.
However I think @uluQulu features are extremely impressive. We definitely should merge them into the code base maybe really setting the default to false instead of true, trying to make the users aware of it.

Do you guys think there will be an easy way to add the "use one account to do the crawling and another one for the interactions"? I mean starting with a non proxy thing were people simply can set up a second account if they want to.
I honestly think that for most people the risks @sionking mentioned are mostly not relevant at this point, sure we should consider them and help people be "as safe as possible", but I guess a lot of people simply want the stuff to work.

I mean if Instagram, at one point really wants to get rid of all accounts using bots, their platform will look like a ghost town in the wild west, unfortunately... So I guess they first want to know how many of the accounts there, surely, use bots which is why they track all the data.

@sionking @uluQulu Thank you very much for all the hard work and the great discussions.
I just love all the features you guys come up with and the amount of work you put into this.
I'm honoured to have people like you really like this tool and contribute to it. Thank you

@uluQulu Let's merge this asap, just making sure that as little of this is "forced" so that people like @sionking are able to use them in their way and don't need to use them in order to use InstaPy.

Best
Tim

uluQulu commented:
It's a great pleasure @timgrossmann,
thank you, @sionking.

@xX7 pagination is no any fixed count, a while back I've taken thorough scroller tests run for `like_by_tags` feature, and I found out ranges in [1 to 20) based on the request frequency.  
And specifically for the dialog boxes, it averagely is ~10 and is not fixed.
_What from your point interested me about AJAX was that it made a sense of where `graphql` is in use which would be so helpful to know._
Well, the pagination size is not the case and won't be a case. Cos using `graphql` is simply noticable, because there is a _direct navigation_, etc.
Although it would be great to conduct a research on the viability of the `graphql` access (_e.g. why it is public, did it have to be public_) and find out the full scope of the capabilities in long distance.
_And yes there are thousands of third-parties who rely on these data scraping functionalities and if one will not work, there will always be another solution._  
So I mean you can use `graphql` for today and yet keep chilling in full comfort. 🍍
Also @xX7, I have already mentioned my concerns about @sionking's approach of additional account+proxy, so anything further will be duplicate. go ahead and use it for yourself or find a stable proxy service with free servers available worldwide for every country which also give a good speed.
Only finding that service, you can think of realizing your suggestion to the general code base.
And yes if you give that proxy finding job into the hands of the user, read my previous comment in there to see get the answer.



@timgrossmann, sure it won't be that hard, it can be a called a twins strategy- each new account entered can have a twins of it.
The thing is having a good storage of that information kept in safe file after making the twins of the newly entered account. So after a twins account is made for the new account, in the next logins it should always use that twins, and not make new ones.
And retire that twins when it does not work correctly :D
It has drawbacks, e.g. newly made accounts generally get less limits of data access (_scraping also may have limits, etc._).
But in any case all of them can be done in a headless mode without the notice of the users to provide seamless activity and also adding a second inner-instance of InstaPy class to do the data acquisition job (_in any case, will take a share from performance_).
>BTW I like ghost towns if it is gonna be like in Casper's. xD

What I like to say guys is to do everything as it have to be and always follow the limits and never exceed them at production environment.

Sure @timgrossmann from latest posts I saw, there are tons of people waiting for these features
And there are another group of guys who want the features which will be available with the help of this toolset.
_E.g._ there is a feature I prepared back then in #1732 which results to follow people WHO do not follow you already..
_It was available for `follow_user_followers` feature only but after these tools, it will be so easy to implement that functionality throughout entire InstaPy- available to all features._

But I will probably will not be able to implement them. Cos, my free days are ended, it's up to you now.

Best regards.

timgrossmann commented:
@xX7 @sionking @uluQulu Alright, let's merge this.

@sionking I hope that's ok for you in the current state. Hopefully you can easily make the adjustments you mentioned to make it work like you want to.

Thank you all for the great work, looking forward to more of this kind.

uluQulu commented:
It's has always been a pleasure to add some piece to this great suit, @timgrossmann.
BTW while on the road, I thought about the safety considerations raised by @sionking (_twins user + proxy_) and have an idea to get it work (_at least I have thought it could work_).
But I need some crucial feedback from @sionking since it's been a long time he is doing using combo.
General questions are:
+ Can you explain a bit about your usage (in technical details)?
+ How reliable & stable it is?
Personal questions are:
+ Since it is being run by Chrome, do you think IG can collect some footprint (_e.g. MAC adress which Chrome can surely access_ and can share with anybody)
+ Any other footprint? (_e.g. sites like Google gets your location data and shares it with other services- I have faced this problem before_)

_If there is ultimate safe to use proxy (_or anything else?_), it is ready to implement with the idea I've thought which  will cover all of the situations regardless of geolocation._
>It is a very important case. Cos in ANY little flaw (_even outside considerations_) in the this particular implementation can be highly serious.
Also, I don't mean a simple usage through some tunnel, rather, it must be considered to handle a huge traffic in clearly non-human means which is gonna be clearly detectable activity.

@sionking you can write to my personal email, too. 😄
>If any of you guys also have some ideas to share in this context, please do it.

condomfighters commented:
Thank you for the amazing work @uluQulu!
